flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/sink/v2/FilesCommitter.java

```diff
@@ -0,0 +1,224 @@
+    }
+
+    commitResult(dataFilesNumStored, deleteFilesNumStored, store);
+  }
+
+  @Override
+  public void close() throws Exception {
+    if (tableLoader != null) {
+      tableLoader.close();
+    }
+
+    if (workerPool != null) {
+      workerPool.shutdown();
+    }
+  }
+
+  private void commitResult(int dataFilesNum, int deleteFilesNum,
+                            Collection<CommitRequest<FilesCommittable>> committableCollection) {
+    int totalFiles = dataFilesNum + deleteFilesNum;
+
+    continuousEmptyCheckpoints = totalFiles == 0 ? continuousEmptyCheckpoints + 1 : 0;
+    if (totalFiles != 0) {
+      if (replacePartitions) {
+        replacePartitions(committableCollection, deleteFilesNum);
+      } else {
+        commitDeltaTxn(committableCollection, dataFilesNum, deleteFilesNum);
+      }
+
+      continuousEmptyCheckpoints = 0;
+    }
+  }
+
+  private void replacePartitions(Collection<CommitRequest<FilesCommittable>> writeResults, int deleteFilesNum) {
+    // Partition overwrite does not support delete files.
+    Preconditions.checkState(deleteFilesNum == 0, "Cannot overwrite partitions with delete files.");
+
+    // Commit the overwrite transaction.
+    ReplacePartitions dynamicOverwrite = table.newReplacePartitions().scanManifestsWith(workerPool);
+
+    int numFiles = 0;
+    String flinkJobId = "";
+    for (CommitRequest<FilesCommittable> writeResult : writeResults) {
+      WriteResult result = writeResult.getCommittable().committable();
+      Preconditions.checkState(result.referencedDataFiles().length == 0, "Should have no referenced data files.");
+      flinkJobId = writeResult.getCommittable().jobID();
+      numFiles += result.dataFiles().length;
+      Arrays.stream(result.dataFiles()).forEach(dynamicOverwrite::addFile);
+    }
+
+    commitOperation(dynamicOverwrite, numFiles, 0, "dynamic partition overwrite", flinkJobId);
+  }
+
+  private void commitDeltaTxn(Collection<CommitRequest<FilesCommittable>> writeResults, int dataFilesNum,
+                              int deleteFilesNum) {
+    if (deleteFilesNum == 0) {
+      // To be compatible with iceberg format V1.
+      AppendFiles appendFiles = table.newAppend().scanManifestsWith(workerPool);
+      String flinkJobId = "";
+      for (CommitRequest<FilesCommittable> commitRequest : writeResults) {
+        try {
+          WriteResult result = commitRequest.getCommittable().committable();
+          Preconditions.checkState(
+              result.referencedDataFiles().length == 0,
+              "Should have no referenced data files.");
+
+          Arrays.stream(result.dataFiles()).forEach(appendFiles::appendFile);
+          flinkJobId = commitRequest.getCommittable().jobID();
+        } catch (Exception e) {
+          commitRequest.signalFailedWithUnknownReason(e);
+        }
+      }
+
+      commitOperation(appendFiles, dataFilesNum, 0, "append", flinkJobId);
+      writeResults.forEach(CommitRequest::signalAlreadyCommitted);
```

Comment:
Based on the Flink documentation:
```
        /**
         * Signals that a committable is skipped as it was committed already in a previous run.
         * Using this method is optional but eases bookkeeping and debugging. It also serves as a
         * code documentation for the branches dealing with recovery.
         */
```

You know more about Flink than me, but this javadoc tells me that we should only call this when we skipped the commit, because it was committed in a previous run. Here it seems to be we are using it as signalling that the commit was successful. Or simply, I just misunderstand something.

Reply:
A commit must be idempotent: If some failure occurs in Flink during commit phase, Flink will restart from previous checkpoint and re-attempt to commit all committables. Thus, some or all committables may have already been committed. These Committer.CommitRequests must not change the external system and implementers are asked to signal Committer.CommitRequest.signalAlreadyCommitted().

I call it mainly to avoid double commits.

Reply:
I have tested the following sequence (slightly modified TestIcebergInputFormats.testFilterExp):
```java
[..]
    helper = new TestHelper(conf, tables, location.toString(), SCHEMA, SPEC, fileFormat, temp);
[..]
    helper.createTable();

    List<Record> expectedRecords = helper.generateRandomRecords(2, 0L);
    expectedRecords.get(0).set(2, "2020-03-20");
    expectedRecords.get(1).set(2, "2020-03-20");

    DataFile dataFile1 = helper.writeFile(Row.of("2020-03-20", 0), expectedRecords);
    DataFile dataFile2 = helper.writeFile(Row.of("2020-03-21", 0), helper.generateRandomRecords(2, 0L));
    helper.appendToTable(dataFile1, dataFile2); // This creates a transaction and adds the data files to it using 'table.newAppend()'
    helper.appendToTable(dataFile1, dataFile2);
```

The resulting table contained the expected records twice. So the Iceberg table commit is not idempotent, adding the same file twice will duplicate the record. For me this means that we have to make sure that the `FilesCommitter.commit` is idempotent ourselves.

Reading only the FlinkSink documentation, I think that calling `commitRequest.signalAlreadyCommitted` does not guarantee that the commit operation will not be called again. I think we can not rely on any of the Flink variables / objects to know if a commit happened or not. I think we have to deduce that from reading the Iceberg table data / metadata for the current snapshot, as our only reliable source of information.

@hililiwei: Does this make sense? Do I miss something?

Thanks,
Peter

Reply:
>So the Iceberg table commit is not idempotent

We are of one mind on this point.

When flink recovers from a  pre-checkpoint, it will retrieve the state data of the checkpoint(When a CommitterOperator receives a CommittableMessage, it stores it in the state.), which contain all `FilesCommittable`, regardless of whether the `FilesCommittable` was committed at the last checkpoint. When the current checkpoint fires, it excludes `FilesCommittable` marked as committed. So, if we don't mark it as committed after a `FilesCommittable` commit, it will be committed again.

Reply:
I am afraid, that we can not base our decisions on what Flink tells us at this point. It might be that the commit is successful, but the system fails, and we do not mark the commit as finished, or we mark it, but the state is restored from an older backup. So even if we call the method, and it does more than documented, we still can not be sure that we do not see this particular commit again in case of a retry.

Again, it is only based on my limited understanding.

Thanks for the patience,
Peter

Reply:
Yes, what you said is right. If we have committed it, but haven't had time to mark it as committed, and the system fails, it will still re-commit when it recover from checkpoint.

Reply:
Thank you very much for your question. It made me realize that I seemed to be relying too much on Flink's own fault-tolerant mechanism. We should go back to using `max-checkpoint-id`.
Multiple `FilesCommittable` are committed simultaneously, but marking as committed is serial. It is not atomic with the data submission action, which leads to the above problems. As you said above,  we have to deduce that from reading the Iceberg table metadata.
When the commit is successful, We need to record `max-checkpoint-id` in the table. When the FilesCommittable re-try , we should check whether its checkpoint id is greater than the max.
Besides, we don't need `signalAlreadyCommitted` anymore.
So let me rewrite this logic.

Thank you.